File: python/vierrataleai/ui/terminal.py
import asyncio
from datetime import datetime
from rich.console import Console
from rich.markdown import Markdown
from .branding import Branding


class Terminal:
    def __init__(self):
        self.console = Console()

    @staticmethod
    def _time():
        return datetime.now().strftime("%H:%M")

    def print_user(self, text: str) -> None:
        self.console.file.write("\r\x1b[K")
        self.console.print()
        self.console.print(f"  [bold #7c3aed]You[/bold #7c3aed]  [dim]{self._time()}[/dim]")
        self.console.print(f"  {text}")
        self.console.print()

    def print_ai_start(self) -> None:
        self.console.print()
        self.console.print(f"  [bold #7c3aed]{Branding.AI_PROMPT}[/bold #7c3aed]  [dim]{self._time()}[/dim]")
        self._prefix = True
        self._line = ""

    def begin_thinking(self) -> None:
        self._thinking = True
        self.console.print(
            "  [#06b6d4]│[/#06b6d4] [dim]· · ·[/dim]", end="", highlight=False
        )

    def end_thinking(self) -> None:
        self.console.file.write("\r\x1b[2K")
        self._prefix = True
        self.console.print("  [#06b6d4]│[/#06b6d4] ", end="", highlight=False)
        self._prefix = False

    def print_ai_chunk(self, text: str) -> None:
        if getattr(self, "_thinking", False):
            self.end_thinking()
            self._thinking = False
        if getattr(self, "_prefix", False):
            self.console.print("  [#06b6d4]│[/#06b6d4] ", end="", highlight=False)
            self._prefix = False
        self.console.print(text, end="", highlight=False)
        self._line += text
        if len(self._line) > 100:
            self._prefix = True
            self._line = ""

    def print_ai_end(self) -> None:
        self.console.print()
        self.console.print()

    def print_error(self, msg: str) -> None:
        self.console.print(f"  [bold red]✖[/bold red] {msg}")

    def print_info(self, msg: str) -> None:
        self.console.print(f"  [dim]{msg}[/dim]")

    def print_success(self, msg: str) -> None:
        self.console.print(f"  [bold green]✔[/bold green] {msg}")

    def print_warning(self, msg: str) -> None:
        self.console.print(f"  [bold yellow]⚠[/bold yellow] {msg}")

    def print_search_results(self, results) -> None:
        self.console.print("  [bold yellow]Web Search Results[/bold yellow]")
        for i, r in enumerate(results, 1):
            self.console.print(
                f"  [bold yellow]{i}.[/bold yellow] [bold]{r['title']}[/bold]"
            )
            self.console.print(f"     [dim]{r['url']}[/dim]")
            if r.get("snippet"):
                self.console.print(f"     {r['snippet'][:120]}")

    async def show_progress(self, steps, work) -> None:
        frames = ["▁", "▂", "▃", "▄", "▅", "▆", "▇", "█"]
        i = 0

        def render():
            nonlocal i
            i += 1
            stage = steps[(i // 8) % len(steps)]
            bar = frames[i % len(frames)]
            fill = "=" * (1 + (i % 18)) + ">" + " " * (18 - (i % 18))
            self.console.file.write(
                f"\r  [{bar}] [dim]{stage}[/dim] [[#06b6d4]{fill}[/#06b6d4]]"
            )
            self.console.file.flush()

        timer = asyncio.get_event_loop().call_later(0, lambda: _tick())

        def _tick():
            render()
            timer = asyncio.get_event_loop().call_later(0.1, _tick)

        try:
            await work()
        finally:
            timer.cancel()
            self.console.file.write("\r\x1b[2K")
            self.console.print(f"  [bold green]✔[/bold green] [dim]Ready[/dim]")
            self.console.print()

    def print_markdown(self, text: str) -> None:
        self.console.print(Markdown(text))

    def clear(self) -> None:
        self.console.clear()

File: nodejs/src/ui/terminal.js
import { Branding } from './branding.js';

const C = Branding.colors;

export const Terminal = {
  divider(char = '─', length = 50) {
    return `${C.dim}${char.repeat(length)}${C.reset}`;
  },

  printUser(text) {
    const time = new Date().toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
    process.stdout.write('\r\x1b[K');
    console.log();
    console.log(`  ${C.bold}${C.primary}You${C.reset}  ${C.dim}${time}${C.reset}`);
    console.log(`  ${text}`);
    console.log();
  },

  printAIStart() {
    const time = new Date().toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
    console.log();
    console.log(`  ${C.bold}${C.accent}${Branding.AI_PROMPT}${C.reset}  ${C.dim}${time}${C.reset}`);
    this.curLine = '';
    this._printPrefix();
  },

  _printPrefix() {
    process.stdout.write(`  ${C.accent}│${C.reset} `);
  },

  beginThinking() {
    process.stdout.write(`${C.dim}· · ·${C.reset}`);
  },

  endThinking() {
    process.stdout.write('\r\x1b[2K');
    this._printPrefix();
  },

  printAIChunk(text) {
    this._thinking = this._thinking || false;
    if (this._thinking) {
      this.endThinking();
      this._thinking = false;
    }
    this.curLine = (this.curLine || '') + text;
    process.stdout.write(text);
    this._maybeWrap();
  },

  _maybeWrap() {
    if (this.curLine && this.curLine.length > 100) {
      const words = this.curLine.split(' ');
      if (words.length > 1) {
        const lastWord = words[words.length - 1];
        const rest = this.curLine.slice(0, this.curLine.length - lastWord.length).trimEnd();
        process.stdout.write(`\n${C.accent}│${C.reset} `);
        this.curLine = lastWord;
      }
    }
  },

  printAIEnd() {
    process.stdout.write('\n');
    console.log();
  },

  printError(msg) {
    console.log(`  ${C.error}✖${C.reset} ${msg}`);
  },

  printInfo(msg) {
    console.log(`  ${C.dim}${msg}${C.reset}`);
  },

  printSuccess(msg) {
    console.log(`  ${C.success}✔${C.reset} ${msg}`);
  },

  printWarning(msg) {
    console.log(`  ${C.warning}⚠${C.reset} ${msg}`);
  },

  printSearchResults(results) {
    console.log(`  ${C.warning}Web Search Results${C.reset}`);
    results.forEach((r, i) => {
      console.log(`  ${C.warning}${i + 1}.${C.reset} ${C.bold}${r.title}${C.reset}`);
      console.log(`     ${C.dim}${r.url}${C.reset}`);
      if (r.snippet) console.log(`     ${r.snippet.slice(0, 120)}`);
    });
  },

  async showProgress(steps, work) {
    // Hidden real progress: the work() task runs silently while we paint a
    // fake animated progress line so the install feels alive.
    const frames = ['▁', '▂', '▃', '▄', '▅', '▆', '▇', '█'];
    let i = 0;
    const timer = setInterval(() => {
      i++;
      const stage = steps[Math.floor((i / 8) % steps.length)];
      const bar = frames[i % frames.length];
      const fill = '='.repeat(1 + (i % 18)) + '>' + ' '.repeat(18 - (i % 18));
      process.stdout.write(`\r  ${C.accent}${bar}${C.reset} ${C.dim}${stage}${C.reset} [${C.accent}${fill}${C.reset}]`);
    }, 100);
    try {
      await work();
    } finally {
      clearInterval(timer);
      process.stdout.write('\r\x1b[2K');
      console.log(`  ${C.success}✔${C.reset} ${C.dim}Ready${C.reset}`);
      console.log();
    }
  },

  clear() {
    process.stdout.write('\x1b[2J\x1b[H');
  },

  formatMarkdown(text) {
    return text
      .replace(/\*\*(.*?)\*\*/g, `${C.bold}$1${C.reset}`)
      .replace(/`([^`]+)`/g, `${C.accent}\`$1\`${C.reset}`)
      .replace(/```(\w*)\n([\s\S]*?)```/g, `${C.dim}┌─${C.reset}\n$2${C.dim}└─────${C.reset}`);
  },
};
